File: 2015/Day3/Day3.py
santa_location = [0,0]
houses_visited = {tuple(santa_location)}

with open("input.txt", 'r') as input_file:
	while True:
		line = input_file.readline()

		if not line:
			break

		for char in line:
			if char == '^':
				santa_location[0] += 1
			elif char == 'v':
				santa_location[0] -= 1
			elif char == '>':
				santa_location[1] += 1
			elif char == '<':
				santa_location[1] -= 1

			houses_visited.add(tuple(santa_location))

print('Number of houses visited:', len(houses_visited))

File: 2015/Day3/Day3.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <array>
#include <set>

using namespace std;

void addToVisitedHouses(const char& c, array<int, 2>& location, set<array<int, 2>>& housesVisited);

int main(){

	array<int, 2> santaLocation = {0, 0};
	array<int, 2> roboSantaLocation = {0, 0};
	set<array<int, 2>> housesVisited;

	string line;
	ifstream inputFile ("input.txt");

	int counter = 0;

  	while (getline(inputFile, line)){
  		for(const char& c : line){
  			if (counter % 2 == 0){
  				addToVisitedHouses(c, santaLocation, housesVisited);
  			}
  			else{
  				addToVisitedHouses(c, roboSantaLocation, housesVisited);
  			}
  			counter++;
  		}
  	}

  	cout << "Number of houses visited: " << housesVisited.size() << endl;

	return 0;
}

void addToVisitedHouses(const char& c, array<int, 2>& location, set<array<int, 2>>& housesVisited){
	if (c == '^'){
		location[0]++;
	}
	else if (c == 'v'){
		location[0]--;
	}
	else if (c == '>'){
		location[1]++;
	}
	else if (c == '<'){
		location[1]--;
	}

	housesVisited.insert(location);
}
